### PR TITLE
[CARBONDATA-3849] pushdown array_contains filter to carbon for array of primitive types

### PR DESCRIPTION
**Pull request by ajantha-bhat**
### Why is this PR needed?
Currently `array_contains()` UDF is not pushed down to carbon. So, carbon has to scan all the rows for query having this UDF. scanning all the rows reduces the query performance. Hence need to pushdown. 

 ### What changes were proposed in this PR?
push down `array_contains()` for **all the array of primitive type** as `equalsTo` filter. Having as equals to filter, we can break the scanning of elements in array once found.
 

 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3050/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1330/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1334/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3054/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3055/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1335/

File: integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestCompactionComplexType.scala [hunk 1]
```diff
@@ -47,6 +47,33 @@ class TestCompactionComplexType extends QueryTest with BeforeAndAfterAll {
...
 
+  test("complex issue") {
+    sql("drop table if exists complex1")
+    sql("create table complex1 (arr array<String>) stored as carbondata")
+    sql("insert into complex1 select array('as') union all " +
+        "select array('sd','df','gh') union all " +
+        "select array('rt','ew','rtyu','jk','sder') union all " +
+        "select array('ghsf','dbv','fg','ty') union all " +
+        "select array('hjsd','fggb','nhj','sd','asd')")
+
```
Comment (Indhumathi27) [inline integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestCompactionComplexType.scala:58]: Please add test scenario with data as null. array(null)

Comment (ajantha-bhat) [reply integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestCompactionComplexType.scala:58]: This is WIP temp, cannot merge this poc code. Why review?

File: core/src/main/java/org/apache/carbondata/core/scan/complextypes/ComplexQueryType.java [hunk 2]
```diff
@@ -67,4 +67,18 @@ private DimensionColumnPage getDecodedDimensionPage(DimensionColumnPage[][] dime
...
   }
+
+  /**
+   * Method will copy the block chunk holder data and return the cloned value.
+   * This method is also used by child.
+   */
+  protected byte[] copyBlockDataChunkWithoutClone(DimensionRawColumnChunk[] rawColumnChunks,
+      DimensionColumnPage[][] dimensionColumnPages, int rowNumber, int pageNumber) {
+    byte[] data =
+        getDecodedDimensionPage(dimensionColumnPages, rawColumnChunks[columnIndex], pageNumber)
```
Comment (QiangCai) [inline core/src/main/java/org/apache/carbondata/core/scan/complextypes/ComplexQueryType.java:78]: how about to cache the page, it will not require to decode for each row again.

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java [hunk 3]
```diff
@@ -222,49 +224,90 @@ public BitSetGroup applyFilter(RawBlockletColumnChunks rawBlockletColumnChunks,
...
+    if (isDimensionPresentInCurrentBlock.length == 1 && isDimensionPresentInCurrentBlock[0]) {
+      // fill default value here
+      DimColumnResolvedFilterInfo dimColumnEvaluatorInfo = dimColEvaluatorInfoList.get(0);
+      // if filter dimension is not present in the current add its default value
+      if (dimColumnEvaluatorInfo.getDimension().getDataType().isComplexType()) {
+        for (int i = 0; i < pageNumbers; i++) {
+          BitSet set = new BitSet(numberOfRows[i]);
+          RowIntf row = new RowImpl();
+          for (int index = 0; index < numberOfRows[i]; index++) {
+            ArrayQueryType complexType =
```
Comment (QiangCai) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java:237]: move to the outside of for loop

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java [hunk 4]
```diff
@@ -222,49 +224,90 @@ public BitSetGroup applyFilter(RawBlockletColumnChunks rawBlockletColumnChunks,
...
+                (ArrayQueryType) complexDimensionInfoMap.get(dimensionChunkIndex[i]);
+            int[] numberOfChild = complexType
+                .getNumberOfChild(rawBlockletColumnChunks.getDimensionRawColumnChunks(), null,
```
Comment (QiangCai) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java:240]: how about to get all numbers of the child once

**Comment by QiangCai** [comment]
if the query has only one simple filter(without and/or), maybe we can try to push down "limit" to filter.
So the filter will not require to read all values of all rows.

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1336/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3056/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1353/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3074/

File: core/src/main/java/org/apache/carbondata/core/scan/complextypes/ComplexQueryType.java (hunk 2 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/scan/complextypes/ComplexQueryType.java]: I have debugged, cache is already there. the argument of this method, `DimensionColumnPage[][] dimensionColumnPages` itself is a cache based on column index.

go inside `ComplexQueryType#getDecodedDimensionPage` to see it.

Also observed that only once decodeColumnPage called for that page, reset it is using from cache only.

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/scan/complextypes/ComplexQueryType.java]: In BlockletScannedResult, dimensionColumnPages[][]

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1408/

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java (hunk 3 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java]: done

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3132/

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java (hunk 4 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java]: done

**Comment by ajantha-bhat** [comment]
@QiangCai : We need spark changes support to push down limit to carbonara. So, I think it cannot be done here as we use open source spark.

I want to implement array_contains pushdown for all the primitive type not just string type. I will finish it today.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3133/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1409/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3139/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1415/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3140/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1416/

File: integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonFilters.scala [hunk 5]
```diff
@@ -152,13 +152,25 @@ object CarbonFilters {
     }
 
     def getCarbonExpression(name: String) = {
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonFilters.scala:154]: in 'createFilter' method,  convert CarbonArrayContains filter to ArrayContainsExpression

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/CarbonLateDecodeStrategy.scala [hunk 6]
```diff
@@ -517,7 +518,8 @@ private[sql] class CarbonLateDecodeStrategy extends SparkStrategy {
       val supportBatch =
         supportBatchedDataSource(relation.relation.sqlContext,
           updateRequestedColumns) && extraRdd.getOrElse((null, true))._2
-      if (!vectorPushRowFilters && !supportBatch && !implicitExisted) {
+      if (!vectorPushRowFilters && !supportBatch && !implicitExisted && filterSet.nonEmpty &&
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/CarbonLateDecodeStrategy.scala:521]: This is for count(*) with array_contains() query.  Here they were reverting back the array_contains(). so avoided it.

**Comment by ajantha-bhat** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3145/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1421/

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java [hunk 7]
```diff
@@ -222,49 +228,103 @@ public BitSetGroup applyFilter(RawBlockletColumnChunks rawBlockletColumnChunks,
...
-      BitSet set = new BitSet(numberOfRows[i]);
-      RowIntf row = new RowImpl();
-      BitSet prvBitset = null;
-      // if bitset pipe line is enabled then use rowid from previous bitset
-      // otherwise use older flow
-      if (!useBitsetPipeLine ||
-          null == rawBlockletColumnChunks.getBitSetGroup() ||
-          null == bitSetGroup.getBitSet(i) ||
-          rawBlockletColumnChunks.getBitSetGroup().getBitSet(i).isEmpty()) {
+    if (isDimensionPresentInCurrentBlock.length == 1 && isDimensionPresentInCurrentBlock[0]
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java:231]: @QiangCai : can you please tell me, why new expression is required ? why equalTo is not enough ?

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/CarbonLateDecodeStrategy.scala [hunk 8]
```diff
@@ -679,18 +681,20 @@ private[sql] class CarbonLateDecodeStrategy extends SparkStrategy {
     // In case of ComplexType dataTypes no filters should be pushed down. IsNotNull is being
     // explicitly added by spark and pushed. That also has to be handled and pushed back to
     // Spark for handling.
-    val predicatesWithoutComplex = predicates.filter(predicate =>
+    // allow array_contains() push down
+    val filteredPredicates = predicates.filter(predicate =>
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/CarbonLateDecodeStrategy.scala:685]: ok

**Comment by ajantha-bhat** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1484/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3211/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3217/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1490/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/CarbonLateDecodeStrategy.scala [hunk 9]
```diff
@@ -865,6 +869,27 @@ private[sql] class CarbonLateDecodeStrategy extends SparkStrategy {
         Some(CarbonContainsWith(c))
       case c@Literal(v, t) if (v == null) =>
         Some(FalseExpr())
+      case c@ArrayContains(a: Attribute, Literal(v, t)) =>
+        a.dataType match {
+          case arrayType: ArrayType =>
+            arrayType.elementType match {
+              case StringType => Some(sources.EqualTo(a.name, v))
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/CarbonLateDecodeStrategy.scala:877]: I want reuse existing equalsTo code, I don't see any advantage of making new expression

File: integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonFilters.scala (hunk 5 above)
Comment (ajantha-bhat) [reply integration/spark/src/main/scala/org/apache/spark/sql/optimizer/CarbonFilters.scala:154]: I want reuse existing equalsTo code, I don't see any advantage of making new expression

**Comment by ajantha-bhat** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3392/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1652/

File: integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestArrayContainsPushDown.scala [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+
+    checkExistence(sql(" explain select * from complex1 where array_contains(arr,'sd')"),
+      true,
+      "PushedFilters: [*EqualTo(arr,sd)]")
+
+    checkExistence(sql(" explain select count(*) from complex1 where array_contains(arr,'sd')"),
+      true,
+      "PushedFilters: [*EqualTo(arr,sd)]")
+
+    checkAnswer(sql(" select * from complex1 where array_contains(arr,'sd')"),
```
Comment (QiangCai) [inline integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestArrayContainsPushDown.scala:57]: can you add a test case that likes the below query?

select * from complex1 where arr[0] = 'sd'

can we push down this filter too?

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/CarbonLateDecodeStrategy.scala [hunk 11]
```diff
@@ -865,7 +870,33 @@ private[sql] class CarbonLateDecodeStrategy extends SparkStrategy {
         Some(CarbonContainsWith(c))
       case c@Literal(v, t) if (v == null) =>
         Some(FalseExpr())
-      case others => None
+      case c@ArrayContains(a: Attribute, Literal(v, t)) =>
+        a.dataType match {
+          case arrayType: ArrayType =>
+            arrayType.elementType match {
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/CarbonLateDecodeStrategy.scala:876]: how about extract the match code block to a method: isPrimitiveDataType and move it into a util class?

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (QiangCai) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RowLevelFilterExecuterImpl.java:231]: it will be hard to read the code after we add more if condition

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/CarbonLateDecodeStrategy.scala (hunk 11 above)
Comment (ajantha-bhat) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/strategy/CarbonLateDecodeStrategy.scala:876]: ok. moved

File: integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestArrayContainsPushDown.scala (hunk 10 above)
Comment (ajantha-bhat) [reply integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestArrayContainsPushDown.scala:57]: Currently carbon doesn't support pushdown of arr[0] = 'sd', because this pushdown is based on array index. 
Need a separate handling for this. yet to analyze the changes.

Comment (ajantha-bhat) [reply integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestArrayContainsPushDown.scala:57]: This PR is only for UDF pushdown

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1690/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3432/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3440/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1698/

**Comment by QiangCai** [comment]
LGTM
